### PR TITLE
Collapse option not working for the the BO category tree

### LINKED ISSUE #23558: "Expand all" button not working in the back office
#### Java-script error when trying to close categories tree
When trying to close a category folder, the folder does not close. An error will be displayed.
This issue can be seen in module ps_facetedsearch 

#### Steps to reproduce the behavior:

1. Go to module ps_facetedsearch
2. Click on button **Add a new template**
3. In the Categories option click the button "Expand All"
4. Try to close an category folder

**Error**
Uncaught TypeError: elem is undefined

**File causing error**
 /backoffice/themes/default/js/tree.js Line 255

#### Additional information

* PrestaShop version: 1.7.7.2
* ps_facetedsearch 
Version 3.5.0 > not working
Version 3.7.0 > not working
* Browsers
Chrome > not working
Firefox > not working

### PR DESCRIPTION
**Pull request by ghost**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | 1.7.7.x
| Description?      |  Resolve issue "collapse" option not working for the the BO category tree (old theme) Affected pages: Faceted search, other pages and modules which use the old theme tree version
| Type?             | bug fix 
| Category?         | BO 
| BC breaks?        | 
| Deprecations?     | 
| Fixed ticket?     | Fixes #23558.
| How to test?      | Displays an error when closing a (sub)category folder in the category tree.
| Possible impacts? | category tree in module Faceted search, other pages and modules which uses the old theme category tree.

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/23625)

### CODE REVIEW

File: admin-dev/themes/default/js/tree.js [hunk 1]
```diff
@@ -213,13 +210,13 @@ Tree.prototype =
...
 				'ajax-tab.php',
-				{controller:'AdminProducts',token:currentToken,action:'getCategoryTree',type:idTree,fullTree:1,selected:selected, inputName:name,useCheckBox:useCheckBox},
+				{ controller:'AdminProducts', token:currentToken, action:'getCategoryTree', type:idTree, fullTree:1, selected:selected, inputName:name, useCheckBox:useCheckBox },
 				function(content)
 				{
 					$('#' + idTree).html(content);
 					organizeTree();
 					$('#' + idTree).tree('init');
-					that.$element.find("label.tree-toggler").each(
+					this.$element.find("label.tree-toggler").each(
```
Comment (NeOMakinG) [inline admin-dev/themes/default/js/tree.js]: What's the reference of `this` there ? It looks like it's scope is the anonymous function

**Comment by NeOMakinG** [comment]
I asked a question only on one this, but every `$(this)` scope looks strange to me

File: admin-dev/themes/default/js/tree.js (hunk 1 above)
Comment (ghost) [reply admin-dev/themes/default/js/tree.js]: This is a mistake it should be the JQuery version $(this).

The issue is that the current code uses variables for $(this)  --> `var that = $(this);` and `var thatOne = $(this);`
However these variables are not passed in the functions `function(content)` which causes the error.

File: admin-dev/themes/default/js/tree.js [hunk 2]
```diff
@@ -213,13 +210,13 @@ Tree.prototype =
...
 				'ajax-tab.php',
-				{controller:'AdminProducts',token:currentToken,action:'getCategoryTree',type:idTree,fullTree:1,selected:selected, inputName:name,useCheckBox:useCheckBox},
+				{ controller:'AdminProducts', token:currentToken, action:'getCategoryTree', type:idTree, fullTree:1, selected:selected, inputName:name, useCheckBox:useCheckBox },
 				function(content)
 				{
 					$('#' + idTree).html(content);
 					organizeTree();
 					$('#' + idTree).tree('init');
-					that.$element.find("label.tree-toggler").each(
+					$(this).$element.find("label.tree-toggler").each(
```
Comment (sowbiba) [inline admin-dev/themes/default/js/tree.js]: hard for me to see here what $(this) refers to

File: admin-dev/themes/default/js/tree.js (hunk 1 above)
Comment (NeOMakinG) [reply admin-dev/themes/default/js/tree.js]: that wasn't accessible ? It looks like it's in the upperscope, it should be accessible ! strange thing there

File: admin-dev/themes/default/js/tree.js (hunk 2 above)
Comment (NeOMakinG) [reply admin-dev/themes/default/js/tree.js]: It looks like it refers to `"label.tree-toggler, .icon-folder-close, .icon-folder-open"` but the actual behavior refers to the tree, I don't know which one is right, I think we need to checkout @crezzur branch in order to review this, overwise you should maybe write a good description to explain it a bit more :)

Comment (ghost) [reply admin-dev/themes/default/js/tree.js]: Not sure what the issue is guys? the element $(this) is a jQuery element which holds the data that was clicked.

As you can see the current code already use the **element $(this)** but puts it into a variable **that** and **thatOne** .
In the current version the variable cannot be accessed from the function. Just try it and you will see i am correct.

**Review by NeOMakinG — DISMISSED** [review]
Tested it and every changed line was using wrong scoped reference, LGTM to me ;)

**Review by PierreRambaud — DISMISSED** [review]

**Comment by khouloudbelguith** [comment]
Hi @crezzur,

Thanks for your fix.
This PR fixes a regression in PS1.7.7, you need to target the `1.7.7.x` branch.

Thank you!

**Comment by crezzur** [comment]
> 
> 
> Hi @crezzur,
> 
> Thanks for your fix.
> This PR fixes a regression in PS1.7.7, you need to target the `1.7.7.x` branch.
> 
> Thank you!

Added the 1.7.7.x branch

**Comment by Progi1984** [comment]
@crezzur No your are targeting `develop` branch and not the `1.7.7.x` branch

**Comment by crezzur** [comment]
> 
> 
> @crezzur No your are targeting `develop` branch and not the `1.7.7.x` branch

I had added the 1.7.7.x branch at start but it was removed. I will reverse these changes now.
**Changes**
PierreRambaud removed the 1.7.7.x label 5 days ago
prestonBot added the develop label 5 days ago

**Comment by Progi1984** [comment]
@crezzur I think you misunderstood me : your PR targets the `develop` branch :
image

**Comment by crezzur** [comment]
@Progi1984  No idea how to change it

**Comment by NeOMakinG** [comment]
@crezzur I think you need to change the target branch to 1.7.7.x, and then rebase your branch on the 1.7.7.x branch, if the rebase is too hard, we can just open a new PR or cherry pick the commit on a new PR

**Comment by crezzur** [comment]
@NeOMakinG Did i do it correct now?

**Comment by NeOMakinG** [comment]
> @NeOMakinG Did i do it correct now?

You just changed the base branch, you need to rebase it into 1.7.7.x now

**Comment by PierreRambaud** [comment]
The PR has been rebased on the 1.7.7.x branch ;)

**Review by Progi1984 — DISMISSED** [review]

**Review by NeOMakinG — DISMISSED** [review]

**Review by matthieu-rolland — DISMISSED** [review]

**Comment by khouloudbelguith** [comment]
Hi @crezzur,

I have an error in the console when we click on the Expand button.
Steps to reproduce the issue:
1. Go to BO > Module Manager page and search for the ps_facetedsearch module
2. Click on configure 
3. Add a new template
4. In the `Categories used for this template:` section > Click on `Expand`
5. See an error in the console 
image

**Without PR** => ok
https://drive.google.com/file/d/1lgXB93jAf_VQERS-2nEJ-CDZWvyfHZW4/view?usp=sharing

Also when the multistore context is enabled, with this PR, the tree issue is fixed in the `Choose shop association` section
image
With PR => OK in different pages (facetedsearch, attributes, features, Customer Group,..) only in the `Add new shop page`.
Steps to reproduce the issue:
1. Go to BO > Shop Parameters > General page and enable multistore
2. Go to BO > Advanced Parameters > Multistore page 
3. Click on "Add new Shop"
4. Click on `Collapse All`
5. Click on `Check All`
6. Click on `Expand All`
7. Try to close a subcategory folder => NOK

https://drive.google.com/file/d/1o_Ew35RLOhKdigh4Ufaq1ketDNAfjgjJ/view?usp=sharing

This issue should be fixed in this PR?

Thank you!

**Comment by ghost** [comment]
> 
> 
> Hi @crezzur,
> 
> I have an error in the console when we click on the Expand button.
> Steps to reproduce the issue:
> 
>     1. Go to BO > Module Manager page and search for the ps_facetedsearch module
> 
>     2. Click on configure
> 
>     3. Add a new template
> 
>     4. In the `Categories used for this template:` section > Click on `Expand`
> 
>     5. See an error in the console
>        image
> 
> 
> **Without PR** => ok
> https://drive.google.com/file/d/1lgXB93jAf_VQERS-2nEJ-CDZWvyfHZW4/view?usp=sharing
> 
> Also when the multistore context is enabled, with this PR, the tree issue is fixed in the `Choose shop association` section
> image
> With PR => OK in different pages (facetedsearch, attributes, features, Customer Group,..) only in the `Add new shop page`.
> Steps to reproduce the issue:
> 
>     1. Go to BO > Shop Parameters > General page and enable multistore
> 
>     2. Go to BO > Advanced Parameters > Multistore page
> 
>     3. Click on "Add new Shop"
> 
>     4. Click on `Collapse All`
> 
>     5. Click on `Check All`
> 
>     6. Click on `Expand All`
> 
>     7. Try to close a subcategory folder => NOK
> 
> 
> https://drive.google.com/file/d/1o_Ew35RLOhKdigh4Ufaq1ketDNAfjgjJ/view?usp=sharing
> 
> This issue should be fixed in this PR?
> 
> Thank you!

Issue resolved by adding $('#' + idTree) to select the correct tree to expand on request.

File: admin-dev/themes/default/js/tree.js [hunk 3]
```diff
@@ -213,13 +210,13 @@ Tree.prototype =
 
 			$.get(
 				'ajax-tab.php',
-				{controller:'AdminProducts',token:currentToken,action:'getCategoryTree',type:idTree,fullTree:1,selected:selected, inputName:name,useCheckBox:useCheckBox},
+				{ controller:'AdminProducts', token:currentToken, action:'getCategoryTree', type:idTree, fullTree:1, selected:selected, inputName:name, useCheckBox:useCheckBox },
 				function(content)
 				{
 					$('#' + idTree).html(content);
```
Comment (sowbiba) [inline admin-dev/themes/default/js/tree.js]: can be improved by putting `$('#' + idTree)` in a variable. But it's not blocking. Thanks @crezzur ;)

**Review by sowbiba — DISMISSED** [review]

File: admin-dev/themes/default/js/tree.js (hunk 3 above)
Comment (ghost) [reply admin-dev/themes/default/js/tree.js]: Added a new commit where the `$('#' + idTree)` is stored in the variable called  `targetTree`

**Review by sowbiba — DISMISSED** [review]
👌🏾

**Review by NeOMakinG — DISMISSED** [review]

**Review by Progi1984 — DISMISSED** [review]

**Comment by khouloudbelguith** [comment]
Hi @crezzur,

The first issue is fixed but the second problem still exists.
Steps to reproduce the issue:

1. Go to BO > Shop Parameters > General page 
2. Enable Multistore context
3. Go to BO > Advanced Parameters > Multistore page 
4. Click on "Add new shop"
5. In the "Associated categories" section, click on `Collapse All`
6. Click on `Check All`
7. Click on `Expand All`
8. Try to close a folder
9. See error: Cannot be closed

I attached a screen record
https://drive.google.com/file/d/1XfrDGN6BgR_LmCS8ALnpO0fDeUEVR7VO/view?usp=sharing

Thank you!

**Comment by ghost** [comment]
@khouloudbelguith
The error is showing instantly when you click on the **expand all** button. The reason it is displayed in the **Add shop** treeview only is because this treeview has the **search option** activated.

The error is caused by the function call `organizeTree(); `(tree.js line 218).
Looking into this function it looks like its a dead part of code.

inside the function `organizeTree()` a call to the function `getCategoryById()` is made. Looking into the function `getCategoryById()` the HTML attribute `name="id_parent"` is searched. 

Searching into the whole folder `STORE\BACKFOFFICE\themes\default\` and more specific `STORE\BACKFOFFICE\themes\default\template\helpers\tree\` for the HTML attribute `name="id_parent"` returns negative.

concluding this i am guessing the functions organizeTree() and getCategoryById() are leftovers from something what is already removed. If you come to the same conclusion i will add a pull request to remove those functions.

**Comment by khouloudbelguith** [comment]
Hi @crezzur,

Thanks for your feedback.
So this issue of creating a shop will be fixed in another PR?

Thank you!

**Comment by ghost** [comment]
@khouloudbelguith added the changes to this pull request.

**Review by NeOMakinG — APPROVED** [review]

**Review by sowbiba — APPROVED** [review]

**Comment by khouloudbelguith** [comment]
Hi @crezzur,

It is ok :heavy_check_mark: 
https://drive.google.com/file/d/1IDkA6nqb1P4OPc22qNTl32c-QXBYcRYD/view?usp=sharing

Thank you!

**Pull request merged**